daemon/settings.c

```diff
@@ -1158,8 +1173,43 @@ appsettings_encode(const appsettings_t *self, GKeyFile *file)
 static void
 appsettings_rethink(appsettings_t *self)
 {
-    /* Assign current value -> masking is re-applied */
-    appsettings_set_granted(self, self->ast_granted);
+    log_info("%s(uid=%d): rethink",
+             appsettings_appname(self),
+             appsettings_uid(self));
+
+    int permission_change = appsettings_update_permissions(self);
+
+    const stringset_t *permissions = appsettings_get_permissions(self);
+    const stringset_t *granted = appsettings_get_granted(self);
+
+    if( appsettings_update_autogrant(self, appsettings_get_allowlisted(self)) ) {
+        /* Autogrant config changed: choose all or nothing */
+        if( appsettings_get_allowed(self) != APP_ALLOWED_NEVER ) {
+            appsettings_update_allowed(self, APP_ALLOWED_UNSET);
+            granted = permissions;
+        }
+    }
+    else {
+        /* Apply autogrant policy */
+        switch( appsettings_get_autogrant(self) ) {
+        case APP_GRANT_ALWAYS:
+            /* Keep in sync with application requirements */
+            granted = permissions;
+            break;
+        case APP_GRANT_LAUNCH:
+            /* Sync if new permissions are required */
+            if( permission_change > 0 )
+                granted = permissions;
```

Comment:
This works a bit differently than it used though. It grants all permissions, not just the added ones. I.e. if some have been denied, then those will be also granted again.

Reply:
Almost missed this...

Reply:
Should now add only freshly added permissions